lib/improver/nowcasting/accumulation.py

```diff
@@ -0,0 +1,214 @@
+accumulations from shorter intervals.
+"""
+import warnings
+import numpy as np
+
+import iris
+from improver.utilities.cube_manipulation import expand_bounds
+from improver.utilities.cube_units import (enforce_coordinate_units_and_dtypes,
+                                           enforce_diagnostic_units_and_dtypes)
+
+
+class Accumulation:
+    """
+    Class to calculate precipitation accumulations from radar rates fields
+    provided at discrete time intervals. The plugin will calculate
+    accumulations between each pair of rates fields provided. These will be
+    used to construct the accumulation period requested when possible, and
+    cubes of this desired period are then returned.
+    """
+    def __init__(self, accumulation_units='m', accumulation_period=None):
+        """
+        Initialise the plugin.
+
+        Args:
+            accumulation_units (str):
+                The physical units in which the accumulation should be
+                returned. The default is metres.
+            accumulation_period (int):
+                The desired accumulation period in seconds. This period
+                must be evenly divisible by the time intervals of the input
+                cubes. The default is None, in which case an accumulation is
+                calculated across the span of time covered by the input rates
+                cubes.
+        """
+        self.accumulation_units = accumulation_units
+        self.accumulation_period = accumulation_period
+
+    def __repr__(self):
+        """Represent the plugin instance as a string."""
+        result = ('<Accumulation: accumulation_units={}, '
+                  'accumulation_period={}>')
+        return result.format(self.accumulation_units, self.accumulation_period)
+
+    @staticmethod
+    def sort_cubes_by_time(cubes):
+        """
+        Sort cubes in time ascending order (from earliest time to latest time).
+
+        Args:
+            cubes (iris.cube.CubeList):
+                A cubelist containing input precipitation rate cubes.
+        Returns:
+            cubes (iris.cube.CubeList):
+                The cubelist in ascending time order.
+            times (list):
+                A list of the validity times of the precipitation rate cubes in
+                integer seconds since 1970-01-01 00:00:00.
+        """
+        times = np.array([cube.coord('time').points[0] for cube in cubes])
+        time_sorted = np.argsort(times)
+        times = times[time_sorted]
+        cubes = iris.cube.CubeList(np.array(cubes)[time_sorted])
+        return cubes, times
+
+    def get_period_sets(self, time_interval, cubes):
+        """
+        Return sub-sets of the input cube list that will be used to construct
+        the accumulation period specified by self.accumulation_period.
+
+        Args:
+            time_interval (int):
+                The time interval between the input precipitation rate cubes in
+                seconds.
```

Comment:
I don't agree with this complaint. I have to calculate the time difference somewhere and I have to check the intervals are regularly spaced. I can do this in a standalone function, in process, or in this function. I have already had to access the time information in order to sort the cubes, so I return this to process to avoid having to access it again, and from there it is a simple call to calculate and check the intervals. I could pass that information (times) into get_period_sets instead, but that is still not the object metadata. What you suggest requires that I access the time coordinate on every cube a second time in this function.

As I am working within a plugin, where the behaviour is sequential and controlled I don't really see the issue with the current approach.

Reply:
OK.  In which case this method should be private, please.

Reply:
Done.